Support for socket options to get/set TOS/TTL info

larseggert opened the pull request:
Add support for getting/setting IPv4 type-of-service (TOS) and time-to-live (TTL) information, and IPv6 traffic class (TC) and hop count information.

Depends on https://github.com/rust-lang/libc/pull/3450, which was recently merged.

(This is my first PR to nix, and I'm also reasonably new to Rust. Please let me know if I missed anything I should have changed; am happy to rework this.)

SteveLauC commented:
Hi! Thanks for your interest in contributing to Nix! :)

A libc with that patch hasn't been released yet, so we have to wait.

And a CHANGELOG entry is needed, please see [CONTRIBUTING.md](https://github.com/nix-rust/nix/blob/master/CONTRIBUTING.md#pull-requests) on how to add one.

larseggert commented:
@SteveLauC Thanks! I will convert this to a draft PR until libc has been release. Will also update CHANGELOG.

asomers reviewed (commented):
Interesting feature.  Would it be possible to add a test?

SteveLauC commented:
Hi @larseggert, we now use the libc dependency from git, so you don't need to patch it yourself, rebasing your branch should make things work

larseggert commented:
@asomers I added a test. Note that there is a `FIXME` in there - `cmsg_space!` returns too little space when called with the option types. I can't figure out why that is.

@SteveLauC I added a CHANGELOG entry.

SteveLauC commented on src/sys/socket/mod.rs:
```diff
@@ -983,6 +1015,55 @@ impl ControlMessageOwned {
+            ))]
+            #[cfg(feature = "net")]
+            (libc::IPPROTO_IP, libc::IP_RECVTOS) => {
+                let tos = unsafe { ptr::read_unaligned(p as *const libc::c_int) };
```
I realized that all these pointer type castings are using the `as` keyword, we should replace them with `.cast()` in a separate PR

larseggert replied:
Should I do this in this PR for the additions I made, or leave as is?

SteveLauC replied:
Leaving them as-is should be ok, we will clear them in another PR

larseggert commented on test/sys/test_socket.rs:
```diff
@@ -2828,3 +2828,87 @@ fn test_icmp_protocol() {
+        let mut buf = [0u8; 8];
+        let mut iovec = [IoSliceMut::new(&mut buf)];
+
+        // FIXME: When passing in IpTos and IpTtl, cmsg_space returns too little space?
+        // let mut space = cmsg_space!(IpTos, IpTtl);
+        let mut space = cmsg_space!(libc::c_int, libc::c_int);
```
@asomers I'm trying to look into this, but I am new to Rust. What I have figured out is that `mem::size_of::< nix::sys::socket::sockopt::IpTos>()` is `0`, and that causes the issue, because `cmsg_space!` allocates too little space. What I don't understand is _why_ it is `0`, given that I have (tried to) make it a `libc::c_int`...

larseggert replied:
Also `test_recv_ipv4pktinfo` also allocates via `let mut space = cmsg_space!(libc::in_pktinfo)` and not via `let mut space = cmsg_space!(nix::sys::socket::sockopt::Ipv4PacketInfo)`, so I guess that's the way to do it?

asomers replied:
Here's the answer: you were trying to calculate the cmsg_space of something unrelated.  Instead of `nix::sys::socket::sockopt::IpTos`, you really wanted `nix::sys::socket::ControlMessageOwned::IpTos`.  The former really does have a size of 0.  And you can't pass the latter to `cmsg_space!`, because `mem::size_of` doesn't work on enum variants.  Does that explanation change anything about the way you would write your test?

larseggert commented on test/sys/test_socket.rs:
```diff
@@ -2828,3 +2828,87 @@ fn test_icmp_protocol() {
+            .flags
+            .intersects(MsgFlags::MSG_TRUNC | MsgFlags::MSG_CTRUNC));
+
+        let mut found = 0;
```
Will do.

larseggert commented on test/sys/test_socket.rs:
```diff
@@ -2828,3 +2828,87 @@ fn test_icmp_protocol() {
+        ),
+        None => return,
+    };
+    let receive = socket(
```
I went with how the other `AF_INET` tests that go over loopback do things. I tried using `socketpair` with `AF_INET`, but I don't think that's supported?

asomers commented on src/sys/socket/mod.rs:
```diff
@@ -803,6 +803,38 @@ pub enum ControlMessageOwned {
+        target_os = "freebsd",
+    ))]
+    #[cfg(feature = "net")]
+    IpTos(libc::c_int),
```
Does this need to be `u8` on linux_android?

larseggert replied:
I don't think so.

larseggert replied:
Actually, hang on. The tests pass but my actual app is not.

SteveLauC commented on src/sys/socket/mod.rs:
```diff
@@ -803,6 +803,38 @@ pub enum ControlMessageOwned {
+        target_os = "freebsd",
+    ))]
+    #[cfg(feature = "net")]
+    Ipv6HopLimit(libc::c_int),
```
This variant is gated with
```rs
    #[cfg(any(linux_android, freebsdlike, apple_targets, target_os = "haiku"))]
```

 in `ControlMessage`, seems that we should also gate it in this way here?

SteveLauC commented:
Hi, sorry for the late response, I think I will take another look at this PR tomorrow:)

SteveLauC commented on src/sys/socket/sockopt.rs:
```diff
@@ -1015,6 +1039,50 @@ sockopt_impl!(
+    libc::IP_RECVTTL,
+    bool
+);
+#[cfg(any(apple_targets, linux_android, target_os = "freebsd",))]
+#[cfg(feature = "net")]
+sockopt_impl!(
+    /// Retrieve the current time-to-live field for every
+    /// IPv4 packet received on this socket.
+    IpTtl,
+    Both,
+    libc::IPPROTO_IP,
+    libc::IP_TTL,
+    libc::c_int
+);
```
There is an `Ipv4Ttl` above this, and they seem to be the same thing?

larseggert replied:
Yes, that was there already, but the `Ipv4Ttl` name is unfortunate, because it's not obvious that this is the socket option that should be used for per-packet CMSG use. What do you prefer I do?

SteveLauC replied:
Ahh, they are the same thing under the hood, but for different usages, then this is a really unfortunate naming, do you mind give them some documents on their typical use cases and how they should be properly used, some examples would be great!

SteveLauC commented on src/sys/socket/sockopt.rs:
```diff
@@ -1015,6 +1039,50 @@ sockopt_impl!(
     libc::IP_TTL,
     libc::c_int
 );
+#[cfg(any(apple_targets, linux_android))]
+#[cfg(feature = "net")]
+sockopt_impl!(
+    /// Retrieve the current time-to-live field for every
+    /// IPv4 packet received on this socket.
+    IpRecvTtl,
+    Both,
+    libc::IPPROTO_IP,
+    libc::IP_RECVTTL,
+    bool
+);
```
From the [man](https://man.freebsd.org/cgi/man.cgi?query=ip&apropos=0&sektion=4&manpath=FreeBSD+14.0-RELEASE&arch=default&format=html) page, this option is also available on FreeBSD.

larseggert replied:
`libc` doesn't seem to export it: https://github.com/nix-rust/nix/pull/2203/checks?check_run_id=20299209049

SteveLauC replied:
I [tried to add these 2 constants to the libc crate](https://github.com/rust-lang/libc/pull/3543), though I have no idea about the CI failure...

SteveLauC replied:
You can also adjust our libc git dependency and enable this option for FreeBSD as that libc PR is merged

SteveLauC commented on src/sys/socket/sockopt.rs:
```diff
@@ -1015,6 +1039,50 @@ sockopt_impl!(
+    libc::IP_TTL,
+    libc::c_int
+);
+#[cfg(any(apple_targets, linux_android))]
+#[cfg(feature = "net")]
+sockopt_impl!(
+    /// Retrieve the current hop limit field for every
+    /// IPv6 packet received on this socket.
+    Ipv6RecvHopLimit,
+    Both,
+    libc::IPPROTO_IPV6,
+    libc::IPV6_RECVHOPLIMIT,
+    bool
+);
```
Even though this option is not present in the corresponding `ip6` man page, there is [a bug report](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=197256), which seems to indicate that FreeBSD has this option supported

larseggert replied:
`libc` doesn't seem to export it: https://github.com/nix-rust/nix/pull/2203/checks?check_run_id=20299209049

SteveLauC commented on src/sys/socket/mod.rs:
```diff
@@ -982,6 +1020,53 @@ impl ControlMessageOwned {
+                let tclass = unsafe { ptr::read_unaligned(p as *const libc::c_int) };
+                ControlMessageOwned::Ipv6TClass(tclass)
+            },
+            #[cfg(any(apple_targets))]
+            #[cfg(feature = "net")]
+            (libc::IPPROTO_IP, libc::IP_RECVTTL) => {
+                let ttl = unsafe { ptr::read_unaligned(p as *const libc::c_int) };
+                ControlMessageOwned::IpTtl(ttl)
+            },
+            #[cfg(linux_android)]
+            #[cfg(feature = "net")]
+            (libc::IPPROTO_IP, libc::IP_TTL) => {
+                let ttl = unsafe { ptr::read_unaligned(p as *const libc::c_int) };
+                ControlMessageOwned::IpTtl(ttl)
+            },
```
Do we miss a FreeBSD impl here?

SteveLauC commented on test/sys/test_sockopt.rs:
We have a test for `Ipv4Ttl` (`test_ttl_opts()`), which should be updated.

larseggert replied:
Updated in which way? Those tests are for setting the TTL with a a sockopt for all packets, not via CMSG per packet.

SteveLauC replied:
Then let's leave it unchanged:)

larseggert replied:
I don't think so: https://github.com/nix-rust/nix/actions/runs/7463986033/job/20309921408?pr=2203#step:4:122

SteveLauC replied:
The CI failed due to the reason that `Ipv6Ttl` is not enabled on apple targets, I added it in #2287, once it is merged, we can enable that test for apple targets.

SteveLauC reviewed (changes requested):
Hi, I have left some comments:)

And, since `IpDontFrag` is also supported in this PR, it will be good if we can have a test for this option.

-----

> And, since `IpDontFrag` is also supported in this PR

This change should be recorded in the `changelog`, something like:

1. Add `IpDontFrag` for Linux/Android/FreeBSD
2. Add `Ipv6DontFrag` for FreeBSD
